#File: SNP analyses from H. armigera RNAseq data
#Project: H. armigera SNP analysis
#Data: file = 'selection'. Data generated by Kumar Saurabh in PoPoolation2
#Aim: To characterise SNPs across the transcriptome of RNAseq pools in
#all genes and differentially expressed candidate genes
#Samples: RNAseq data described in Jones et al. (2015)
#Mol Ecol. Four populations/phenotypes (AY, AY, GR1-3, GR4-6). 
#Updated: 01-03-17

#This analysis is performed on Fsts calculated using the 
#following parameters
#minimum coverage = 8
#pool size = 9
#min MAF = 3
#max coverage = 200

############################################
#ANNOTATED MISSENSE FILES FOR EACH COMPARISON WERE POOLED AND THE
#DUPLICATES REMOVED. VARIANTS GENERATED USING SNPEFF
#USING INDIVIDUAL SAMPLE FILES. ANNOTATIONS COULD ORIGINATE FROM EITHER
#SAMPLE COMPARISON SO EACH SNP OF INTEREST NEEDS TO BE INTERROGATED 
#INDIVIDUALLY
############################################
#
#
#set working directory
setwd("H:/R/popoolation/missense/010317_analysis/")

#load packages
library(ggplot2)
library(gridExtra)

#Objective 1
#calculate Fst in the China and Greece comparisons using
#SNPs identified in gene models only

#File; SNPs = AY_DF_missense_rm dup.tab
#load data
CH = read.table("AY_DF_missense_dup rm.tab", h=TRUE) #81984
head(CH)

#Task1.1
#How many SNPs segregate between the China and Greek populations?
#first analyse China by removing the Greece column
CH$greece <- NULL
#remove windows with 0 value
CH<-CH[!(CH$china=="0"),] #AY = 81548

#PLOT READS VERSUS FST
a1 <- ggplot(CH, aes(china, reads)) + geom_point()
a1

mean(allCH$china) #0.0948

#Objective2: Outlier analysis of the 'global analysis'
#Produce a list of outlier SNPs for the top .05, 1, 5%
#single SNPs, China
quantile(allCH$china, c(.95, .99, .995)) #       95%       99%     99.5% 
#                                               0.2762171 0.4582813 0.5479459 

cSNP_OL5 <- subset(allCH, china >0.2762171, na.rm=TRUE) #4078
cSNP_OL1 <- subset(allCH, china >0.4582813, na.rm=TRUE) #861
cSNP_OL05 <- subset(allCH, china >0.5479459, na.rm=TRUE) #408

#Number of SNPs completely fixed
which(allCH$china==1) #9
range(cSNP_OL05$china) #0.5481147 1.0000000
#write the files
write.csv(as.data.frame(cSNP_OL05),
          file="CHINA_missense_OL05.csv")
write.csv(as.data.frame(cSNP_OL1),
          file="CHINA_missense_OL1.csv")
write.csv(as.data.frame(cSNP_OL5),
          file="CHINA_missense_OL5.csv")

#REPEAT FOR GREECE
#Input the combined greek sample file
greece = read.table("greece_missense_rm dup.tab", h=TRUE)
head(all)

#Task1.1
greece$china <- NULL
#remove windows with 0 value
GR<-greece[!(greece$greece=="0"),] #56437

#calculate the mean Fst value
mean(GR$greece) # 0.05098937

#Objective2: Outlier analysis of the 'global analysis'
#Produce a list of outlier SNPs for the top .05, 1, 5%
#single SNPs, China
quantile(GR$greece, c(.95, .99, .995)) #    95%       99%     99.5% 
#                                             0.1736574 0.3043478 0.3670492  

gSNP_OL5 <- subset(GR, greece >0.1736574, na.rm=TRUE) #1531
gSNP_OL1 <- subset(GR, greece >0.3043478, na.rm=TRUE) #307
gSNP_OL05 <- subset(GR, greece >0.3670492, na.rm=TRUE) #154

#Number of SNPs completely fixed
which(GR$greece==1) #0
#
range(gSNP_OL5$greece) #0.174-0.902

#write the files
write.csv(as.data.frame(gSNP_OL05),
          file="GREECE_missense_OL05.csv")
write.csv(as.data.frame(gSNP_OL1),
          file="GREECE_missense_OL1.csv")
write.csv(as.data.frame(gSNP_OL5),
          file="GREECE_missense_OL5.csv")


#COMMON
common_OL05 <- merge(cSNP_OL05, gSNP_OL05, by=c("scaffold","window")) #4
common_OL1 <- merge(cSNP_OL1, gSNP_OL1, by=c("scaffold","window")) #24
common_OL5 <- merge(cSNP_OL5, gSNP_OL5, by=c("scaffold","window")) #187

write.csv(common_OL05, file="common_missense_OL05.csv", row.names=F)
write.csv(common_OL1, file="common_missense_OL1.csv", row.names=F)
write.csv(common_OL5, file="common_missense_OL5.csv", row.names=F)

########################################################################
#PLOT FST VALUES FOR CHINA AND GREECE AGAINST EACH OTHER FOR COMMON 
commonOL5 = read.csv("common_missense_OL5.csv", h=TRUE)
a1 <- ggplot(commonOL5, aes(china,greece)) +
  geom_point()
a1

########################################################################
#DISTRIBUTION OF FST MISSENSE
########################################################################
#task1.3
#plot the distribution (density plot) of Fst into bins similar to Fig2. 
#in Akey et al. (2015)
a1 <- ggplot(CH, aes(x=china)) +
  geom_histogram(aes(y=..density..), binwidth=0.05, colour="black", fill="white")
a2 <- a1 + xlab("Fst") + ylab("Density") + theme_bw() + xlim(0,1) + ylim(0,15)
a3 <- a2 + theme(panel.grid.minor=element_blank(), panel.grid.major=element_blank())
a3
